### PR TITLE
Safe VocbasePtr

### PR DESCRIPTION
**Pull request by gnusi**
### Scope & Purpose

Make vocbase ref counting safer.

- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [x] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: *(Please link PR)*
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/Cluster/HeartbeatThread.cpp [hunk 1]
```diff
@@ -1358,16 +1358,10 @@ bool HeartbeatThread::handlePlanChangeCoordinator(uint64_t currentPlanVersion) {
...
         TRI_ASSERT(db);
         if (db) {
-          try {
-            dbName = db->name();
-          } catch (...) {
-            db->release();
-            throw;
-          }
-          db->release();
+          dbName = db->name();
```
Comment (MBkkt) [inline arangod/Cluster/HeartbeatThread.cpp:1364]:
```suggestion
          dbName = db->name();
          db = {};
```

File: arangod/VocBase/Methods/Databases.cpp [hunk 2]
```diff
@@ -423,14 +421,13 @@ ErrorCode dropDBCoordinator(DatabaseFeature& df, std::string const& dbName) {
...
   while (++tries <= 6000) {
-    TRI_vocbase_t* vocbase = df.useDatabase(id);
+    auto vocbase = df.useDatabase(id);
 
     if (vocbase == nullptr) {
       // object has vanished
       break;
     }
 
-    vocbase->release();
```
Comment (MBkkt) [inline arangod/VocBase/Methods/Databases.cpp:433]: vocbase = {};

File: arangod/VocBase/Methods/Databases.cpp [hunk 3]
```diff
@@ -401,16 +401,14 @@ arangodb::Result Databases::create(ArangodServer& server,
...
-  TRI_vocbase_t* vocbase = df.useDatabase(dbName);
+  auto vocbase = df.useDatabase(dbName);
 
   if (vocbase == nullptr) {
     return TRI_ERROR_ARANGO_DATABASE_NOT_FOUND;
   }
 
   TRI_voc_tick_t const id = vocbase->id();
 
-  vocbase->release();
```
Comment (MBkkt) [inline arangod/VocBase/Methods/Databases.cpp:412]: vocbase = {};

File: arangod/V8Server/v8-vocbase.cpp [hunk 4]
```diff
@@ -1676,10 +1676,9 @@ static void JS_UseDatabase(v8::FunctionCallbackInfo<v8::Value> const& args) {
   void* orig = v8g->_vocbase;
   TRI_ASSERT(orig != nullptr);
 
-  v8g->_vocbase = vocbase;
-  static_cast<TRI_vocbase_t*>(orig)->release();
```
Comment (jsteemann) [inline arangod/V8Server/v8-vocbase.cpp:1680]: I think this is still required. If we are not calling `release()` on `orig`, the reference count of `orig` will never be counted down again.

File: arangod/Utils/DatabaseGuard.cpp [hunk 5]
```diff
@@ -29,13 +29,13 @@ namespace {
 
 template<typename T>
 TRI_vocbase_t& vocbase(arangodb::DatabaseFeature& feature, T& id) {
```
Comment (jsteemann) [inline arangod/Utils/DatabaseGuard.cpp:31]: I don't really understand this function, although it was available before this PR:
it returns a reference to a `TRI_vocbase_t`, but to do so it increases the vocbase's reference count... And it doesn't even return a pointer.
Can we get rid of the entire function?

Comment (gnusi) [reply arangod/Utils/DatabaseGuard.cpp:31]: Sure

File: arangod/V8Server/v8-vocbase.cpp [hunk 6]
```diff
@@ -1676,10 +1676,9 @@ static void JS_UseDatabase(v8::FunctionCallbackInfo<v8::Value> const& args) {
...
+  v8g->_vocbase = vocbase.get();
```
Comment (MBkkt) [inline arangod/V8Server/v8-vocbase.cpp:1679]:
```suggestion
  auto* prev = static_cast<TRI_vocbase_t*>(v8g->_vocbase);
  v8g->_vocbase = vocbase.release();
  vocbase.get_deleter()(prev);
```

**Review by MBkkt — CHANGES_REQUESTED** [review]

File: arangod/V8Server/V8DealerFeature.cpp [hunk 7]
```diff
@@ -565,17 +565,13 @@ void V8DealerFeature::start() {
...
-      try {
-        context = buildContext(vocbase, nextId());
-        TRI_ASSERT(context != nullptr);
-      } catch (...) {
-        vocbase->release();
-      }
+      V8Context* context = buildContext(vocbase.get(), nextId());
+      TRI_ASSERT(context != nullptr);
+
+      vocbase->release();
```
Comment (MBkkt) [inline arangod/V8Server/V8DealerFeature.cpp:574]: vocbase = {};

Comment (MBkkt) [inline arangod/V8Server/V8DealerFeature.cpp:574]:
```cpp
try {
  V8Context* context = buildContext(vocbase.get(), nextId());
  TRI_ASSERT(context != nullptr);
  vocbase.release();
} catch (...) {
   vocbase = {};
}
```

File: arangod/Cluster/HeartbeatThread.cpp [hunk 8]
```diff
@@ -1387,12 +1387,13 @@ bool HeartbeatThread::handlePlanChangeCoordinator(uint64_t currentPlanVersion) {
...
       auto dbName = info.getName();
-      TRI_vocbase_t* vocbase = databaseFeature.useDatabase(dbName);
+      auto vocbase = databaseFeature.useDatabase(dbName);
       if (vocbase == nullptr) {
         // database does not yet exist, create it now
 
         // create a local database object...
-        Result res = databaseFeature.createDatabase(std::move(info), vocbase);
+        [[maybe_unused]] TRI_vocbase_t* unused{};
+        Result res = databaseFeature.createDatabase(std::move(info), unused);
```
Comment (MBkkt) [inline arangod/Cluster/HeartbeatThread.cpp:1396]:
```cpp
        TRI_vocbase_t* new_database{};
        Result res = databaseFeature.createDatabase(std::move(info), new_database);
        vocbase = new_database
```

Better to accept VocbasePtr& in createDatabase

File: arangod/V8Server/v8-vocbase.cpp (hunk 4 above)
Comment (gnusi) [reply arangod/V8Server/v8-vocbase.cpp:1680]: Doesn't whatever produced by `WrapVocBase(isolate, vocbase)` decrease the count?

File: arangod/V8Server/V8DealerFeature.cpp (hunk 7 above)
Comment (gnusi) [reply arangod/V8Server/V8DealerFeature.cpp:574]: I know, could be addressed in the subsequent PR(s)

File: arangod/RestServer/DatabaseFeature.h [hunk 9]
```diff
@@ -239,43 +242,36 @@ class DatabaseFeature : public ArangodFeature {
...
-  replication::Version _defaultReplicationVersion;
+  bool _defaultWaitForSync{false};
+  bool _ignoreDatafileErrors{false};
+  bool _isInitiallyEmpty{false};
+  bool _checkVersion{false};
+  bool _upgrade{false};
+  // allow extended database names or not
+  bool _extendedNamesForDatabases{false};
+  bool _performIOHeartbeat{true};
+  std::atomic<bool> _started{false};
```
Comment (MBkkt) [inline arangod/RestServer/DatabaseFeature.h:253]: NIT
```suggestion
  std::atomic_bool _started{false};
```

File: arangod/Utils/DatabaseGuard.cpp [hunk 10]
```diff
@@ -24,41 +24,35 @@
...
-
-template<typename T>
-TRI_vocbase_t& vocbase(arangodb::DatabaseFeature& feature, T& id) {
-  auto* vocbase = feature.useDatabase(id);
+namespace arangodb {
 
-  if (!vocbase) {
-    THROW_ARANGO_EXCEPTION(TRI_ERROR_ARANGO_DATABASE_NOT_FOUND);
+void VocbaseReleaser::operator()(TRI_vocbase_t* vocbase) const noexcept {
+  if (vocbase) {
```
Comment (MBkkt) [inline arangod/Utils/DatabaseGuard.cpp:32]: Btw unique_ptr::deleter can be called only explicit or if stored pointer not null

File: arangod/Utils/DatabaseGuard.cpp [hunk 11]
```diff
@@ -24,41 +24,35 @@
...
     THROW_ARANGO_EXCEPTION(TRI_ERROR_ARANGO_DATABASE_NOT_FOUND);
   }
 }
 
 /// @brief create the guard, using a database id
 DatabaseGuard::DatabaseGuard(DatabaseFeature& feature, TRI_voc_tick_t id)
-    : _vocbase(vocbase(feature, id)) {
-  TRI_ASSERT(!_vocbase.isDangling());
+    : _vocbase{feature.useDatabase(id)} {
+  TRI_ASSERT(!_vocbase->isDangling());
```
Comment (MBkkt) [inline arangod/Utils/DatabaseGuard.cpp:49]:
```suggestion
  TRI_ASSERT(_vocbase);
  TRI_ASSERT(!_vocbase->isDangling());
```

File: arangod/Utils/DatabaseGuard.cpp [hunk 12]
```diff
@@ -24,41 +24,35 @@
...
 }
 
 /// @brief create the guard, using a database name
-DatabaseGuard::DatabaseGuard(DatabaseFeature& feature, std::string const& name)
-    : _vocbase(vocbase(feature, name)) {
-  TRI_ASSERT(!_vocbase.isDangling());
+DatabaseGuard::DatabaseGuard(DatabaseFeature& feature, std::string_view name)
+    : _vocbase{feature.useDatabase(name)} {
+  TRI_ASSERT(!_vocbase->isDangling());
```
Comment (MBkkt) [inline arangod/Utils/DatabaseGuard.cpp:55]:
```suggestion
  TRI_ASSERT(_vocbase);
  TRI_ASSERT(!_vocbase->isDangling());
```

File: arangod/V8Server/v8-vocbase.cpp [hunk 13]
```diff
@@ -1673,13 +1673,10 @@ static void JS_UseDatabase(v8::FunctionCallbackInfo<v8::Value> const& args) {
   TRI_ASSERT(!vocbase->isDangling());
 
   // switch databases
-  void* orig = v8g->_vocbase;
+  VocbasePtr orig{std::exchange(v8g->_vocbase, vocbase.release())};
```
Comment (MBkkt) [inline arangod/V8Server/v8-vocbase.cpp:1676]: NIT
```suggestion
  vocbase = std::exchange(v8g->_vocbase, vocbase.release());
```

File: arangod/VocBase/Methods/Databases.cpp [hunk 14]
```diff
@@ -400,20 +400,23 @@ arangodb::Result Databases::create(ArangodServer& server,
 
 namespace {
 ErrorCode dropDBCoordinator(DatabaseFeature& df, std::string const& dbName) {
-  // Arguments are already checked, there is exactly one argument
-  TRI_vocbase_t* vocbase = df.useDatabase(dbName);
+  auto [ci, id] = [&]() -> std::pair<ClusterInfo*, TRI_voc_tick_t> {
```
Comment (MBkkt) [inline arangod/VocBase/Methods/Databases.cpp:403]:
```suggestion
  auto const [ci, id] = [&]() -> std::pair<ClusterInfo*, TRI_voc_tick_t> {
```

**Review by MBkkt — APPROVED** [review]

**Review by jsteemann — APPROVED** [review]
LGTM

File: arangod/V8Server/V8DealerFeature.cpp [hunk 15]
```diff
@@ -565,21 +565,15 @@ void V8DealerFeature::start() {
...
 
-      V8Context* context;
-      try {
-        context = buildContext(vocbase, nextId());
-        TRI_ASSERT(context != nullptr);
-      } catch (...) {
-        vocbase->release();
-      }
+      auto context = buildContext(std::move(vocbase), nextId());
+      TRI_ASSERT(context != nullptr);
```
Comment (MBkkt) [inline arangod/V8Server/V8DealerFeature.cpp:572]:
```suggestion
      auto context = buildContext(vocbase.get(), nextId());
      TRI_ASSERT(context != nullptr);
      vocbase.release();
```

File: arangod/V8Server/V8DealerFeature.cpp [hunk 16]
```diff
@@ -751,45 +745,33 @@ void V8DealerFeature::copyInstallationFiles() {
...
-      return context;
-    } catch (...) {
-      delete context;
-      throw;
-    }
-  } catch (...) {
-    vocbase->release();
-    throw;
-  }
+  return context;
```
Comment (MBkkt) [inline arangod/V8Server/V8DealerFeature.cpp:774]:
```suggestion
  vocbase.release();
  return context;
```

**Pull request merged**